File: scripts/lib_py/firestore_client.py
# Client Firestore partagé côté Python — miroir de scripts/lib_js/firestore-client.js
# (même projet Firebase, mêmes credentials). Init paresseuse : ne se déclenche qu'au
# premier get_db(), jamais à l'import, pour ne pas casser les scripts qui n'ont pas
# besoin de Firestore (ex. gen-maillage.py une fois la niche déjà chargée en mémoire).
import json
import os

_app = None
_db = None


def _load_dotenv_fallback():
    root = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
    env_path = os.path.join(root, '.env')
    if not os.path.exists(env_path):
        return {}
    env = {}
    with open(env_path, encoding='utf-8') as f:
        for line in f:
            if '=' not in line:
                continue
            key, _, value = line.partition('=')
            env[key.strip()] = value.strip()
    return env


_dotenv = _load_dotenv_fallback()


def _get_var(name):
    return os.environ.get(name) or _dotenv.get(name)


def get_db():
    global _app, _db
    if _db is not None:
        return _db
    raw = _get_var('FIREBASE_SERVICE_ACCOUNT_JSON')
    if not raw:
        raise RuntimeError(
            "firestore_client: FIREBASE_SERVICE_ACCOUNT_JSON doit etre defini (env ou .env local) — "
            "voir plan MVP SaaS interne pour la procedure de creation du compte de service Firebase."
        )
    try:
        service_account = json.loads(raw)
    except json.JSONDecodeError as e:
        raise RuntimeError(f"firestore_client: FIREBASE_SERVICE_ACCOUNT_JSON illisible (JSON invalide) : {e}")
    import firebase_admin
    from firebase_admin import credentials, firestore
    if not firebase_admin._apps:
        _app = firebase_admin.initialize_app(credentials.Certificate(service_account))
    else:
        _app = firebase_admin.get_app()
    _db = firestore.client()
    return _db

File: scripts/lib_js/firestore-client.js
// Client Firestore partagé — un seul point d'init pour tout le pipeline JS
// (niche-config.js, state.js, tracking-store.js, dashboard). Pattern repris
// de tonton-maj-v2 (même utilisateur, déjà en production dessus) : credentials
// via FIREBASE_SERVICE_ACCOUNT_JSON (JSON complet du compte de service en une
// variable d'env), pas un fichier sur disque — adapté à un cron GitHub Actions
// ou un déploiement Vercel, aucun des deux n'ayant de disque persistant pour
// un secret fichier. Repli .env local pour le développement, même convention
// que wp-client.js/mistral-client.js.
//
// Init PARESSEUSE (au premier getDb(), jamais au require) : de nombreux
// scripts du repo n'ont besoin ni de près ni de loin de Firestore (ex. tout
// ce qui ne touche qu'à data/maillage/*.json) — les faire échouer au chargement
// faute de credentials casserait des usages qui n'en ont jamais eu besoin.
const fs = require('fs');
const path = require('path');

function loadDotEnvFallback() {
  const envPath = path.join(__dirname, '..', '..', '.env');
  if (!fs.existsSync(envPath)) return {};
  const raw = fs.readFileSync(envPath, 'utf8');
  const env = {};
  raw.split('\n').filter(Boolean).forEach((line) => {
    const i = line.indexOf('=');
    if (i === -1) return;
    env[line.slice(0, i).trim()] = line.slice(i + 1).trim();
  });
  return env;
}

const dotEnv = loadDotEnvFallback();
function getVar(name) {
  return process.env[name] ?? dotEnv[name];
}

let app = null;
let db = null;

function initFirebase() {
  if (app) return app;
  const raw = getVar('FIREBASE_SERVICE_ACCOUNT_JSON');
  if (!raw) {
    throw new Error(
      'firestore-client: FIREBASE_SERVICE_ACCOUNT_JSON doit être défini (env ou .env local) — '
      + "voir plan MVP SaaS interne pour la procédure de création du compte de service Firebase."
    );
  }
  let serviceAccount;
  try {
    serviceAccount = JSON.parse(raw);
  } catch (e) {
    throw new Error(`firestore-client: FIREBASE_SERVICE_ACCOUNT_JSON illisible (JSON invalide) : ${e.message}`);
  }
  // require() différé : évite de charger firebase-admin (et son coût
  // d'import) pour les scripts qui ne finissent jamais par appeler getDb().
  // API modulaire (firebase-admin v13) : `cert`/`initializeApp`/`getApps`
  // sont exportés directement, plus de namespace `admin.credential`/
  // `admin.firestore()` (breaking change par rapport aux versions < v12).
  const admin = require('firebase-admin');
  const { getFirestore } = require('firebase-admin/firestore');
  app = admin.getApps().length ? admin.getApp() : admin.initializeApp({ credential: admin.cert(serviceAccount) });
  db = getFirestore(app);
  return app;
}

function getDb() {
  if (!db) initFirebase();
  return db;
}

module.exports = { getDb, initFirebase };
